### ISSUE TITLE
False positive improper_ctypes warning for `()` used through trait projection

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#66374]
This code:

```rust
pub trait Foo {
    type Ret;
}

impl Foo for u32 {
    type Ret = ();
}

pub extern "C" fn foo() -> <u32 as Foo>::Ret {
}
```

generates this warning on nightly:

```
warning: `extern` fn uses type `()`, which is not FFI-safe
 --> src/lib.rs:9:28
  |
9 | pub extern "C" fn foo() -> <u32 as Foo>::Ret {
  |                            ^^^^^^^^^^^^^^^^^ not FFI-safe
  |
  = note: `#[warn(improper_ctypes)]` on by default
  = help: consider using a struct instead
  = note: tuples have unspecified layout
```

but I believe that this is an false positive because `()` in a return position is indeed FFI safe

**Comment by Centril** [comment]
Duplicate of https://github.com/rust-lang/rust/issues/66202?

**Comment by Centril** [comment]
cc @davidtwco @rkruppe @eddyb

**Comment by davidtwco** [comment]
> Duplicate of #66202?

I think so.

**Comment by rkruppe** [comment]
Agreed, closing as such.
